### What is Python?
Python is a popular programming language. It was created by Guido van Rossum, and released in 1991.

#### It is used for:

1. web development (server-side),
2. software development,
3. mathematics,
4. system scripting.
#### What can Python do?
a. Python can be used on a server to create web applications.

b. Python can be used alongside software to create workflows.

c. Python can connect to database systems. It can also read and modify files.

d. Python can be used to handle big data and perform complex mathematics.

e. Python can be used for rapid prototyping, or for production-ready software development.

#### Why Python?
a. Python works on different platforms (Windows, Mac, Linux, Raspberry Pi, etc).

b. Python has a simple syntax similar to the English language.

c. Python has syntax that allows developers to write programs with fewer lines than some other programming languages.

d. Python runs on an interpreter system, meaning that code can be executed as soon as it is written. This means that prototyping can be very quick.

e. Python can be treated in a procedural way, an object-oriented way or a functional way.

#### Good to know
a. The most recent major version of Python is Python 3, which we shall be using in this tutorial. However, Python 2, although not being updated with anything other than security updates, is still quite popular.

b. In this tutorial Python will be written in a text editor. It is possible to write Python in an Integrated Development Environment, such as Thonny, Pycharm, Netbeans or Eclipse which are particularly useful when managing larger collections of Python files.

#### Python Syntax compared to other programming languages

a. Python was designed for readability, and has some similarities to the English language with influence from mathematics.

b. Python uses new lines to complete a command, as opposed to other programming languages which often use semicolons or parentheses.

c. Python relies on indentation, using whitespace, to define scope; such as the scope of loops, functions and classes. Other programming languages often use curly-brackets for this purpose.

In [1]:
print("Hello,World!")

Hello,World!


### Python Variables.

Variables are containers for storing data values.



In [2]:
x = 5
y = "John"
print(x)
print(y)

5
John


In [3]:
#Variables do not need to be declared with any particular type, and can even change type after they have been set.

x = 4       # x is of type int
x = "Sally" # x is now of type str
print(x)

Sally


### CASTING

If you want to specify the data type of a variable, this can be done with casting.

In [4]:
x = str(3)    # x will be '3'
y = int(3)    # y will be 3
z = float(3)  # z will be 3.0
print(x,y,z,sep=",")

3,3,3.0


### Python - Variable Names

A variable can have a short name (like x and y) or a more descriptive name (age, carname, total_volume). Rules for Python variables:

  1. A variable name must start with a letter or the underscore character
  2. A variable name cannot start with a number
  3. A variable name can only contain alpha-numeric characters and underscores (A-z, 0-9, and _ )
    Variable names are case-sensitive (age, Age and AGE are three different variables)
  4. A variable name cannot be any of the Python keywords.


#### Example for Legal variable names:-

myvar = "Nagarjun"
my_var = "Naga"
_my_var = "Nagarjun"
myVar = "Nagesh"
MYVAR = "Naga_Bhai"
myvar2 = "Nagarjun"

#### Examples for Illegal Variable names:-

2myvar = "John"
my-var = "John"
my var = "John"

In [5]:
# Python Variables - Assign Multiple Values

x, y, z = "Nagarjun", "Vijay", "Arjun"
print(x)
print(y)
print(z)

# Note: Make sure the number of variables matches the number of values, or else you will get an error.

Nagarjun
Vijay
Arjun


In [6]:
# One Value to Multiple Variables

x = y = z = "Nagarjun"
print(x)
print(y)
print(z)

Nagarjun
Nagarjun
Nagarjun


### Unpack a Collection

If you have a collection of values in a list, tuple etc. Python allows you to extract the values into variables. This is called unpacking.

In [7]:
fruits = ["apple", "banana", "cherry"]
x, y, z = fruits
print(x)
print(y)
print(z)

apple
banana
cherry


### Python - Output Variables

In [8]:
# The Python print() function is often used to output variables
x = "Python is awesome"
print(x)
print("-------------------------------------------------------------")

# In the print() function, you output multiple variables, separated by a comma:
x = "Python"
y = "is"
z = "awesome"
print(x, y, z)
print("-------------------------------------------------------------")

# You can also use the + operator to output multiple variables(Space in Values):

x = "Python "
y = "is "
z = "awesome"
print(x + y + z)

Python is awesome
-------------------------------------------------------------
Python is awesome
-------------------------------------------------------------
Python is awesome


In [9]:
x = 5
y = 10
print(x + y)

15


In [10]:
x = 5
y = "John"
print(x + y)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [11]:
x = 5
y = "John"
print(x, y)

5 John


In [12]:
x = "5"
y = "John"
print(x + y)

5John


In [13]:
# Concatenation
x = "Python"
y = "is"
z = "awesome"
print(x + y + z)

Pythonisawesome


In [14]:
a = 50
b = a
print(id(a))
print(id(b))

140721015748936
140721015748936


### Python Data Types

In programming, data type is an important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:

Text Type:	str
Numeric Types:	int, float, complex
Sequence Types:	list, tuple, range
Mapping Type:	dict
Set Types:	set, frozenset
Boolean Type:	bool
Binary Types:	bytes, bytearray, memoryview
None Type:	NoneType

In [15]:
x = 5
print(type(x))

<class 'int'>


### Python Numbers
There are three numeric types in Python:

1. int
2. float
3. complex

In [16]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'float'>
<class 'complex'>


In [17]:
#Int, or integer, is a whole number, positive or negative, without decimals, of unlimited length.
x = 1
y = 35656222554887711
z = -3255522

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'int'>
<class 'int'>


In [18]:
# Float, or "floating point number" is a number, positive or negative, containing one or more decimals.
x = 1.10
y = 1.0
z = -35.59

print(type(x))
print(type(y))
print(type(z))
print("---------------------------------")
# Float can also be scientific numbers with an "e" to indicate the power of 10.
x = 35e3
y = 12E4
z = -87.7e100

print(type(x))
print(type(y))
print(type(z))
print("---------------------------------")

x = 3+5j
y = 5j
z = -5j

print(type(x))
print(type(y))
print(type(z))


<class 'float'>
<class 'float'>
<class 'float'>
---------------------------------
<class 'float'>
<class 'float'>
<class 'float'>
---------------------------------
<class 'complex'>
<class 'complex'>
<class 'complex'>


### Type Conversion
You can convert from one type to another with the int(), float(), and complex() methods:

In [19]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

#convert from int to float:
a = float(x)

#convert from float to int:
b = int(y)

#convert from int to complex:
c = complex(x)

print(a)
print(b) # Lowest Value taken
print(c)

print(type(a))
print(type(b))
print(type(c))


# Note: You cannot convert complex numbers into another number type.

1.0
2
(1+0j)
<class 'float'>
<class 'int'>
<class 'complex'>


### Python Casting

#### Specify a Variable Type
There may be times when you want to specify a type on to a variable. This can be done with casting. Python is an object-orientated language, and as such it uses classes to define data types, including its primitive types.

----> Casting in python is therefore done using constructor functions:

1. int() - constructs an integer number from an integer literal, a float literal (by removing all decimals), or a string literal (providing the string represents a whole number)
2. float() - constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)
3. str() - constructs a string from a wide variety of data types, including strings, integer literals and float literals

In [20]:
# Integers

x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3 and only number in the string will convert to integer.

print(x)
print(y)
print(z)

1
2
3


In [21]:
# Floats

x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

print(x)
print(y)
print(z)
print(w)

1.0
2.8
3.0
4.2


In [22]:
# Strings

x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

print(x)
print(y)
print(z)

s1
2
3.0


### Python Strings

Strings in python are surrounded by either single quotation marks, or double quotation marks.

'hello' is the same as "hello".

You can display a string literal with the print() function:

In [23]:
print("Hello")
print('Hello')

Hello
Hello


In [24]:
# Assigning a string to a variable is done with the variable name followed by an equal sign and the string:

a = "Hello"
print(a)
print("------------------------------------------")
# Multiline Strings

str = '''Twinkle, twinkle, little star,
How I wonder what you are.
Up above the world so high,
Like a diamond in the sky.'''
print(str)
print("------------------------------------------")

Hello
------------------------------------------
Twinkle, twinkle, little star,
How I wonder what you are.
Up above the world so high,
Like a diamond in the sky.
------------------------------------------


### Strings are Arrays
Like many other popular programming languages, strings in Python are arrays of bytes representing unicode characters.
However, Python does not have a character data type, a single character is simply a string with a length of 1.
Square brackets can be used to access elements of the string.

In [25]:
a = "Hello, World!"
print(a[1])

e


In [26]:
#Check String

str = input("Enter the string :")

if "A" in str:
    print("Yes")

else:
    print("Not in String")



Not in String


### Slicing In the String

You can return a range of characters by using the slice syntax.
Specify the start index and the end index, separated by a colon, to return a part of the string.

Syntax:-

         str[start:(End-1)]

         str[start:(End-1):step]

In [27]:
s = "Python Class"
print(s)

Python Class


In [28]:
print(s[7])
print(s[-5])

C
C


In [29]:
print(s[0:6])
print(s[0:-6])

Python
Python


In [30]:
# String Reverse and straight

print(s[::-1])
print(s[::1])

ssalC nohtyP
Python Class


In [31]:
print(s[::-2])
print(s[::2])

saCnhy
Pto ls


In [32]:
b = "Hello, World!"
print(b[-5:-2])

orl


 ### String In-Built Functions


        Method	        Description
       capitalize()	    Converts the first character to upper case

       casefold()	    Converts string into lower case

       center()	        Returns a centered string

       count()	        Returns the number of times a specified value occurs in a string

       encode()	        Returns an encoded version of the string

       endswith()	    Returns true if the string ends with the specified value

       expandtabs()	    Sets the tab size of the string

       find()	        Searches the string for a specified value and returns the position of where it was found

       format()	        Formats specified values in a string

       format_map()	    Formats specified values in a string

       index()	        Searches the string for a specified value and returns the position of where it was found

       isalnum()	    Returns True if all characters in the string are alphanumeric

       isalpha()	    Returns True if all characters in the string are in the alphabet

       isascii()	    Returns True if all characters in the string are ascii characters

       isdecimal()	    Returns True if all characters in the string are decimals

       isdigit()	    Returns True if all characters in the string are digits

       isidentifier()	Returns True if the string is an identifier

       islower()	    Returns True if all characters in the string are lower case

       isnumeric()	    Returns True if all characters in the string are numeric

       isprintable()	Returns True if all characters in the string are printable

       isspace()	    Returns True if all characters in the string are whitespaces

       istitle()	    Returns True if the string follows the rules of a title

       isupper()	    Returns True if all characters in the string are upper case

       join()	        Converts the elements of an iterable into a string

       ljust()	        Returns a left justified version of the string

       lower()	        Converts a string into lower case

       lstrip()	        Returns a left trim version of the string

       maketrans()	    Returns a translation table to be used in translations

       partition()	    Returns a tuple where the string is parted into three parts

       replace()	    Returns a string where a specified value is replaced with a specified value

       rfind()	        Searches the string for a specified value and returns the last position of where it was found

       rindex()	        Searches the string for a specified value and returns the last position of where it was found

       rjust()	        Returns a right justified version of the string

       rpartition()	    Returns a tuple where the string is parted into three parts

       rsplit()	        Splits the string at the specified separator, and returns a list

       rstrip()	        Returns a right trim version of the string

       split()	        Splits the string at the specified separator, and returns a list

       splitlines()	    Splits the string at line breaks and returns a list

       startswith()	    Returns true if the string starts with the specified value

       strip()	        Returns a trimmed version of the string

       swapcase()	    Swaps cases, lower case becomes upper case and vice versa

       title()	        Converts the first character of each word to upper case

       translate()	    Returns a translated string

       upper()	        Converts a string into upper case

       zfill()	        Fills the string with a specified number of 0 values at the beginning



### 1. capitalize()
###### The capitalize() method returns a string where the first character is upper case, and the rest is lower case.
      syntax:-
            string.capitalize()

In [33]:
txt = "python is FUN!"

x = txt.capitalize()

print (x)

Python is fun!


In [34]:
# Nothing is happen if your first character is number
txt = "36 is my age."

x = txt.capitalize()

print (x)

36 is my age.


### 2. casefold()
###### a. The casefold() method returns a string where all the characters are lower case.
###### b. This method is similar to the lower() method, but the casefold() method is stronger, more aggressive, meaning that it will convert more characters into lower case, and will find more matches when comparing two strings and both are converted using the casefold() method.

    Syntax:- string.casefold()



In [35]:
Str = "NAGARJUN Is A King"

print(Str.casefold())

nagarjun is a king


### 3. center()
###### The center() method will center align the string, using a specified character (space is default) as the fill character.
    Syntax:- string.center(length, character)



    Parameter	Description
    length	Required. The length of the returned string
    character	Optional. The character to fill the missing space on each side. Default is " " (space)


In [36]:
Str = "Nagarjun"

x = Str.center(20) # Space is Default

print(x)

      Nagarjun      


In [37]:
Str = "Nagarjun"

x = Str.center(20,"x")

print(x)

xxxxxxNagarjunxxxxxx


### 4. Count()
###### The count() method returns the number of times a specified value appears in the string.
         Syntax:-
            string.count(value, start, end)


      Parameter	Description
       value	Required. A String. The string to value to search for
       start	Optional. An Integer. The position to start the search. Default is 0
       end	Optional. An Integer. The position to end the search. Default is the end of the string

In [38]:
txt = "I love apples, apple are my favorite fruit"

x = txt.count("apple", 10, 24)

print(x)

print("------------------------------------")

txt = "I love apples, apple are my favorite fruit"

x = txt.count("apple")

print(x)

1
------------------------------------
2


### 5. expandtabs()

###### The expandtabs() method sets the tab size to the specified number of whitespaces.

       Syantax:-
               string.expandtabs(tabsize)

      tabsize	Optional. A number specifying the tabsize. Default tabsize is 8



In [39]:
# It is just like putting gap between the letters.
txt = "H\te\tl\tl\to"
x =  txt.expandtabs(2)
print(x)

print("--------------------------------")

txt = "H\te\tl\tl\to"

print(txt)
print(txt.expandtabs())
print(txt.expandtabs(2))
print(txt.expandtabs(4))
print(txt.expandtabs(10))

H e l l o
--------------------------------
H	e	l	l	o
H       e       l       l       o
H e l l o
H   e   l   l   o
H         e         l         l         o


In [40]:
# 6. Use of endswith()---------> string.endswith(value, start, end)

txt = "Hello, welcome to my world."
x = txt.endswith("my world.")
print(x)
print("--------------------------------")

txt = "Hello, welcome to my world."
x = txt.endswith("my world.", 5, 11)
print(x)

True
--------------------------------
False


### 7. find()
1. The find() method finds the first occurrence of the specified value.
2. The find() method returns -1 if the value is not found.
3. The find() method is almost the same as the index() method, the only difference is that the index() method raises an exception if the value is not found. (See example below)

       Syntax:-
               string.find(value, start, end)

In [41]:
txt = "Hello, welcome to my world."
x = txt.find("e")
print(x)
print("----------------------------------")

txt = "Hello, welcome to my world."
x = txt.find("e", 5, 10)
print(x)
print("----------------------------------")

txt = "Hello, welcome to my world."

print(txt.find("q"))
print(txt.index("q"))

1
----------------------------------
8
----------------------------------
-1


ValueError: substring not found

### 8. format()

1. The format() method formats the specified value(s) and insert them inside the string's placeholder.
2. The placeholder is defined using curly brackets: {}. Read more about the placeholders in the Placeholder section below.
3. The format() method returns the formatted string.


#### placeholders

###### The placeholders can be identified using named indexes {price}, numbered indexes {0}, or even empty placeholders {}.

      Syntax:-
            string.format(value1, value2...)


     Formatting Types:-
           Inside the placeholders you can add a formatting type to format the result:

     :<		Left aligns the result (within the available space)
     :>		Right aligns the result (within the available space)
     :^		Center aligns the result (within the available space)
     :=		Places the sign to the left most position
     :+		Use a plus sign to indicate if the result is positive or negative
     :-		Use a minus sign for negative values only
     : 		Use a space to insert an extra space before positive numbers (and a minus sign before negative numbers)
     :,		Use a comma as a thousand separator
     :_		Use a underscore as a thousand separator
     :b		Binary format
     :c		Converts the value into the corresponding unicode character
     :d		Decimal format
     :e		Scientific format, with a lower case e
     :E		Scientific format, with an upper case E
     :f		Fix point number format
     :F		Fix point number format, in uppercase format (show inf and nan as INF and NAN)
     :g		General format
     :G		General format (using a upper case E for scientific notations)
     :o		Octal format
     :x		Hex format, lower case
     :X		Hex format, upper case
     :n		Number format
     :%		Percentage format

In [42]:
txt = "For only {price:.2f} dollars!"
print(txt.format(price = 49))
print("----------------------------")
txt1 = "My name is {fname}, I'm {age}".format(fname = "John", age = 36)
txt2 = "My name is {0}, I'm {1}".format("John",36)
txt3 = "My name is {}, I'm {}".format("John",36)
print(txt1)
print(txt2)
print(txt3)


For only 49.00 dollars!
----------------------------
My name is John, I'm 36
My name is John, I'm 36
My name is John, I'm 36


### 9. index()
1. The index() method finds the first occurrence of the specified value.
2. The index() method raises an exception if the value is not found.
3. The index() method is almost the same as the find() method, the only difference is that the find() method returns -1 if the value is not found. (See example below)

       Syntax:-
               string.index(value, start, end)


In [43]:
txt = "Hello, welcome to my world."
x = txt.index("welcome")
print(x)
print("--------------------------")

txt = "Hello, welcome to my world."
x = txt.index("e")
print(x)
print("--------------------------")

txt = "Hello, welcome to my world."
x = txt.index("e", 5, 10)
print(x)
print("--------------------------")

txt = "Hello, welcome to my world."

print(txt.find("q"))
print(txt.index("q"))

7
--------------------------
1
--------------------------
8
--------------------------
-1


ValueError: substring not found

### 10. isalnum()
1. The isalnum() method returns True if all the characters are alphanumeric, meaning alphabet letter (a-z) and numbers (0-9).
   Example of characters that are not alphanumeric: (space)!#%&? etc.

       Syntax:-
              string.isalnum()



In [44]:
txt = "Company12"
x = txt.isalnum()
print(x)
print("--------------------------")

txt = "Company 12"
x = txt.isalnum()
print(x)

True
--------------------------
False


### 11. isalpha()
1. The isalpha() method returns True if all the characters are alphabet letters (a-z).
   Example of characters that are not alphabet letters: (space)!#%&? etc.

       Syntax:-
              string.isalpha()




In [45]:
txt = "CompanyX"
x = txt.isalpha()
print(x)
print("--------------------------")

txt = "Company10"
x = txt.isalpha()
print(x)

True
--------------------------
False


### 12. isdigit()
1. The isdigit() method returns True if all the characters are digits, otherwise False.
   Exponents, like ², are also considered to be a digit.


         Syntax:-
               string.isdigit()


In [46]:
txt = "50800"
x = txt.isdigit()
print(x)
print("--------------------------")

a = "\u0030" #unicode for 0
b = "\u00B2" #unicode for ²

print(a.isdigit())
print(b.isdigit())

True
--------------------------
True
True


### 13. isidentifier()
1. The isidentifier() method returns True if the string is a valid identifier, otherwise False.
2. A string is considered a valid identifier if it only contains alphanumeric letters (a-z) and (0-9), or underscores (_). A valid identifier cannot start with a number, or contain any spaces.

        Syntax:-
               string.isidentifier()


### 14. islower()
1. The islower() method returns True if all the characters are in lower case, otherwise False.
Numbers, symbols and spaces are not checked, only alphabet characters.

        Syntax:-
               string.islower()


### 15. isspace()
The isspace() method returns True if all the characters in a string are whitespaces, otherwise False.

        Syntax:-
               string.isspace()


### 16. istitle()
1. The istitle() method returns True if all words in a text start with a upper case letter, AND the rest of the word are lower case letters, otherwise False.

2. Symbols and numbers are ignored.

        Syntax:-
               string.istitle()


### 17. isupper()
1. The isupper() method returns True if all the characters are in upper case, otherwise False.
Numbers, symbols and spaces are not checked, only alphabet characters.

        Syntax:-
               string.isupper()


### 18. join()
1. The join() method takes all items in an iterable and joins them into one string.
2. A string must be specified as the separator.

        Syntax:-
               string.join(iterable)
###### Note: When using a dictionary as an iterable, the returned values are the keys, not the values.


### 19. ljust()
The ljust() method will left align the string, using a specified character (space is default) as the fill character.

        Syntax:-
               string.ljust(length, character)



### 20. lower()
1. The lower() method returns a string where all characters are lower case.
Symbols and Numbers are ignored.

        Syntax:-
               string.lower()


### 21. lstrip()
The lstrip() method removes any leading characters (space is the default leading character to remove)

        Syntax:-
               string.lstrip(characters)


### 22. maketrans()
The maketrans() method returns a mapping table that can be used with the translate() method to replace specified characters.

        Syntax:-
               str.maketrans(x, y, z)


    Parameter	Description
      x	Required. If only one parameter is specified, this has to be a dictionary describing how to perform the replace. If two or more parameters are specified, this parameter has to be a string specifying the characters you want to replace.
      y	Optional. A string with the same length as parameter x. Each character in the first parameter will be replaced with the corresponding character in this string.
      z	Optional. A string describing which characters to remove from the original string.


### 23. partition()
The replace() method replaces a specified phrase with another specified phrase.

        Syntax:-
               string.replace(oldvalue, newvalue, count)


### 24. replace()
The replace() method replaces a specified phrase with another specified phrase.

        Syntax:-
               string.replace(oldvalue, newvalue, count)


### 25. split()
1. The split() method splits a string into a list.
2. You can specify the separator, default separator is any whitespace.

        Syntax:-
               string.split(separator, maxsplit)


       Parameter	 Description
       separator	 Optional. Specifies the separator to use when splitting the string. By default any whitespace is a separator
       maxsplit	 Optional. Specifies how many splits to do. Default value is -1, which is "all occurrences"


### 26. startswitch()
The startswith() method returns True if the string starts with the specified value, otherwise False.

        Syntax:-
               string.startswith(value, start, end)


### 27. strip()
1. The strip() method removes any leading, and trailing whitespaces.
Leading means at the beginning of the string, trailing means at the end.
2. You can specify which character(s) to remove, if not, any whitespaces will be removed.

        Syntax:-
               string.strip(characters)


### 28. swapcase()
The swapcase() method returns a string where all the upper case letters are lower case and vice versa.

        Syntax:-
               string.swapcase()


### 29. title()
1. The title() method returns a string where the first character in every word is upper case. Like a header, or a title.
2. If the word contains a number or a symbol, the first letter after that will be converted to upper case.

        Syntax:-
               string.title()

### 30. translate()
1. The translate() method returns a string where some specified characters are replaced with the character described in a dictionary, or in a mapping table.
2. Use the maketrans() method to create a mapping table.
3. If a character is not specified in the dictionary/table, the character will not be replaced.
4. If you use a dictionary, you must use ascii codes instead of characters.

        Syntax:-
               string.translate(table)


### 31. upper()
1. The upper() method returns a string where all characters are in upper case.
2. Symbols and Numbers are ignored.

        Syntax:-
               string.upper()


### 32. zfill()
1. The zfill() method adds zeros (0) at the beginning of the string, until it reaches the specified length.
2. If the value of the len parameter is less than the length of the string, no filling is done.

        Syntax:-
               string.zfill(len)


In [47]:
#Identifier
txt = "Demo"
x = txt.isidentifier()
print(x)
print("--------------------")

a = "MyFolder"
b = "Demo002"
c = "2bring"
d = "my demo"

print(a.isidentifier())
print(b.isidentifier())
print(c.isidentifier())
print(d.isidentifier())

True
--------------------
True
True
False
False


In [48]:
#islower
txt = "hello world!"
x = txt.islower()
print(x)
print("----------------")

a = "Hello world!"
b = "hello 123"
c = "mynameisPeter"

print(a.islower())
print(b.islower())
print(c.islower())

True
----------------
False
True
False


In [49]:
#isspace
txt = "   "
x = txt.isspace()
print(x)
print("---------------")

txt = "   s   "
x = txt.isspace()
print(x)

True
---------------
False


In [50]:
#istitle
txt = "Hello, And Welcome To My World!"
x = txt.istitle()
print(x)
print("--------------------")

a = "HELLO, AND WELCOME TO MY WORLD"
b = "Hello"
c = "22 Names"
d = "This Is %'!?"

print(a.istitle())
print(b.istitle())
print(c.istitle())
print(d.istitle())

True
--------------------
False
True
True
True


In [51]:
#isupper
txt = "THIS IS NOW!"
x = txt.isupper()
print(x)
print("---------------")

a = "Hello World!"
b = "hello 123"
c = "MY NAME IS PETER"

print(a.isupper())
print(b.isupper())
print(c.isupper())

True
---------------
False
False
True


In [52]:
#join
myTuple = ("John", "Peter", "Vicky")
x = "#".join(myTuple)
print(x)
print("--------------------------")

myDict = {"name": "John", "country": "Norway"}
mySeparator = "TEST"

x = mySeparator.join(myDict)

print(x)

#Note: When using a dictionary as an iterable, the returned values are the keys, not the values.

John#Peter#Vicky
--------------------------
nameTESTcountry


In [53]:
#ljust
txt = "banana"
x = txt.ljust(20)
print(x, "is my favorite fruit.")
print("-------------------------------")

txt = "banana"
x = txt.ljust(20, "O")
print(x)

banana               is my favorite fruit.
-------------------------------
bananaOOOOOOOOOOOOOO


In [54]:
#lower
txt = "Hello my FRIENDS"
x = txt.lower()
print(x)
print("-------------------------")



hello my friends
-------------------------


In [55]:
#lstrip
txt = "     banana     "
x = txt.lstrip()
print("of all fruits", x, "is my favorite")
print("----------------------------------")

txt = ",,,,,ssaaww.....banana"
x = txt.lstrip(",.asw")
print(x)

of all fruits banana      is my favorite
----------------------------------
banana


In [56]:
#maketrans
#Create a mapping table, and use it in the translate() method to replace any "S" characters with a "P" character:
txt = "Hello Sam!"
mytable = str.maketrans("S", "P")
print(txt.translate(mytable))
print("-----------------------")
#The maketrans() method returns a mapping table that can be used with the translate() method to replace specified characters.
txt = "Hi Sam!"
x = "mSa"
y = "eJo"
mytable = str.maketrans(x, y)
print(txt.translate(mytable))
print("-----------------------")
#Use a mapping table to replace many characters:
txt = "Good night Sam!"
x = "mSa"
y = "eJo"
z = "odnght"
mytable = str.maketrans(x, y, z)
print(txt.translate(mytable))
print("-----------------------")
#The third parameter in the mapping table describes characters that you want to remove from the string:
txt = "Good night Sam!"
x = "mSa"
y = "eJo"
z = "odnght"
print(str.maketrans(x, y, z))

Hello Pam!
-----------------------
Hi Joe!
-----------------------
G i Joe!
-----------------------
{109: 101, 83: 74, 97: 111, 111: None, 100: None, 110: None, 103: None, 104: None, 116: None}


In [57]:
#partition
#Search for the word "bananas", and return a tuple with three elements:
txt = "I could eat bananas all day"
x = txt.partition("bananas")
print(x)
print("---------------------------------")

#If the specified value is not found, the partition() method returns a tuple containing: 1 - the whole string, 2 - an empty string, 3 - an empty string:
txt = "I could eat bananas all day"
x = txt.partition("apples")
print(x)

('I could eat ', 'bananas', ' all day')
---------------------------------
('I could eat bananas all day', '', '')


In [58]:
#replace
txt = "I like bananas"
x = txt.replace("bananas", "apples")
print(x)
print("--------------------------")

txt = "one one was a race horse, two two was one too."
x = txt.replace("one", "three")
print(x)
print("--------------------------")

txt = "one one was a race horse, two two was one too."
x = txt.replace("one", "three", 2)
print(x)

I like apples
--------------------------
three three was a race horse, two two was three too.
--------------------------
three three was a race horse, two two was one too.


In [59]:
#split
txt = "welcome to the jungle"
x = txt.split()
print(x)
print("--------------------------")

txt = "hello, my name is Peter, I am 26 years old"
x = txt.split(", ")
print(x)
print("--------------------------")

txt = "apple#banana#cherry#orange"
x = txt.split("#")
print(x)
print("--------------------------")

txt = "apple#banana#cherry#orange"
# setting the maxsplit parameter to 1, will return a list with 2 elements!
x = txt.split("#", 1)
print(x)

['welcome', 'to', 'the', 'jungle']
--------------------------
['hello', 'my name is Peter', 'I am 26 years old']
--------------------------
['apple', 'banana', 'cherry', 'orange']
--------------------------
['apple', 'banana#cherry#orange']


In [60]:
#splitlines
txt = "Thank you for the music\nWelcome to the jungle"
x = txt.splitlines()
print(x)
print("--------------------------")
#The splitlines() method splits a string into a list. The splitting is done at line breaks.
txt = "Thank you for the music\nWelcome to the jungle"
x = txt.splitlines(True)
print(x)


['Thank you for the music', 'Welcome to the jungle']
--------------------------
['Thank you for the music\n', 'Welcome to the jungle']


In [61]:
#startswith
txt = "Hello, welcome to my world."
x = txt.startswith("Hello")
print(x)
print("--------------------------")

txt = "Hello, welcome to my world."
x = txt.startswith("wel", 7, 20)
print(x)

True
--------------------------
True


In [62]:
#strip
txt = "     banana     "
x = txt.strip()
print("of all fruits", x, "is my favorite")
print("--------------------------")

txt = ",,,,,rrttgg.....banana....rrr"
x = txt.strip(",.grt")
print(x)

of all fruits banana is my favorite
--------------------------
banana


In [63]:
#swapcase
txt = "Hello My Name Is PETER"
x = txt.swapcase()
print(x)

hELLO mY nAME iS peter


In [64]:
#title
txt = "Welcome to my world"
x = txt.title()
print(x)
print("--------------------------")

txt = "Welcome to my 2nd world"
x = txt.title()
print(x)
print("--------------------------")

txt = "hello b2b2b2 and 3g3g3g"
x = txt.title()
print(x)

Welcome To My World
--------------------------
Welcome To My 2Nd World
--------------------------
Hello B2B2B2 And 3G3G3G


In [65]:
#translate
#use a dictionary with ascii codes to replace 83 (S) with 80 (P):
mydict = {83:  80}
txt = "Hello Sam!"
print(txt.translate(mydict))
print("--------------------------")
#Use a mapping table to replace "S" with "P":
txt = "Hello Sam!"
mytable = str.maketrans("S", "P")
print(txt.translate(mytable))
print("--------------------------")
#Use a mapping table to replace many characters:
txt = "Hi Sam!"
x = "mSa"
y = "eJo"
mytable = str.maketrans(x, y)
print(txt.translate(mytable))
print("--------------------------")
#The third parameter in the mapping table describes characters that you want to remove from the string:
txt = "Good night Sam!"
x = "mSa"
y = "eJo"
z = "odnght"
mytable = str.maketrans(x, y, z)
print(txt.translate(mytable))
print("--------------------------")
#The same example as above, but using a dictionary instead of a mapping table:
txt = "Good night Sam!"
mydict = {109: 101, 83: 74, 97: 111, 111: None, 100: None, 110: None, 103: None, 104: None, 116: None}
print(txt.translate(mydict))

Hello Pam!
--------------------------
Hello Pam!
--------------------------
Hi Joe!
--------------------------
G i Joe!
--------------------------
G i Joe!


In [66]:
#upper
txt = "Hello my friends"
x = txt.upper()
print(x)
print("-----------------------")



HELLO MY FRIENDS
-----------------------


In [67]:
#zfill
txt = "50"
x = txt.zfill(10)
print(x)
print("-----------------------------")

a = "hello"
b = "welcome to the jungle"
c = "10.000"

print(a.zfill(10))
print(b.zfill(10))
print(c.zfill(10))

0000000050
-----------------------------
00000hello
welcome to the jungle
000010.000


In [68]:
#String Concatenation
a = "Hello"
b = "World"
c = a + b
print(c)
print("--------------------")

a = "Hello"
b = "World"
c = a + " " + b
print(c)

HelloWorld
--------------------
Hello World


### Format Strings
1. The format() method allows you to format selected parts of a string.
2. Sometimes there are parts of a text that you do not control, maybe they come from a database, or user input?
3. To control such values, add placeholders (curly brackets {}) in the text, and run the values through the format() method:

### Multiple Values
If you want to use more values, just add more values to the format() method:

### Index Numbers
You can use index numbers (a number inside the curly brackets {0}) to be sure the values are placed in the correct placeholders:

### Named Indexes
You can also use named indexes by entering a name inside the curly brackets {carname}, but then you must use names when you pass the parameter values txt.format(carname = "Ford"):

In [69]:
age = 36
txt = "My name is John, and I am {}"
print(txt.format(age))

My name is John, and I am 36


In [70]:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want {} pieces of item {} for {} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item 567 for 49.95 dollars.


In [71]:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want to pay {2} dollars for {0} pieces of item {1}."
print(myorder.format(quantity, itemno, price))

I want to pay 49.95 dollars for 3 pieces of item 567.


In [72]:
quantity = 3
itemno = 567
price = 49
myorder = "I want {} pieces of item number {} for {:.2f} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item number 567 for 49.00 dollars.


### 1. PYTHON OPERATORS

#### Arithmatic Operators

1. Addition (a+b)

2. Subtraction (a-b)

3. Multiplication (a*b)

4. Divide (a/b)
In this O/P will be QUOTIENT as Float Number.

5. Modulus (a%b)
In this O/P will be Remainder.

6. Floor Division (a//b)
In this O/P will be Quotient as Whole Number.

7. Exponent (a**b)
This means a to the power b.

In [73]:
a = eval(input("Enter the first number :"))
b = eval(input("Enter the second number :"))


print(a+b)

print(a-b)

print(a*b)

print(a/b) #Output will be float number (Quotient)

print(a%b) #Remainder

print(a//b) #While Working with float number division it will take O/P as nearest number to the answer.(Quotient)

print(a**b)

17
7
60
2.4
2
2
248832


In [74]:
a = 12.5
b = 5

print(a//b)

2.0


In [75]:
a = 15.1

b = 5

print(a//b)

3.0


In [76]:
a = -12.5
b = 5

print(a//b)

-3.0


In [77]:
a = 12
b = 5

print(a%b) #Reamainder is the O/P

2


### 2. Relational Operator.

1. Less than (<)

2. Greater than (>)

3. Less than equal to (<=)

4. Greater tha equal to (>=)

5. Equal to (==)

6. Not equal to (!=)

In [78]:
a = int(input("Enter the first number :"))
b = int(input("Enter the second number :"))

print(a<b)

print(a>b)

print(a<=b)

print(a>=b)

print(a==b)

print(a!=b)

False
True
False
True
False
True


### 3. Assignment Operator

1. a/=b ------> a = a/b
2. a+=b ------> a = a+b
3. a-=b ------> a = a-b
4. a*=b ------> a = a*b
5. a%=b ------> a = a%b
6. a//=b ------> a = a//b
7. a&=b ------> a = a&b
8. a^=b ------> a = a^b
9. a>>=b ------> a = a>>b
10. a<<=b -----> a = a<<b

In [79]:
a = 12
b = 4

a/=b
print(a)

3.0


### 4. Logical Operators

A. Logical AND (and)

   A       B      A and B
   0       0         0
   0       1         0
   1       0         0
   1       1         1

     a. One Of the Statement is FALSE(0) Final Statement is FALSE(0)
     b. First Statement is TRUE(1) Then Check the second statement
     c. Both The statements are TRUE(1) then Output is TRUE(1)

B. Logical OR (or)

   A       B      A or B
   0       0        0
   0       1        1
   1       0        1
   1       1        1

     a. One Of the statement is TRUE(1) Then Final Value Is True(1)
     b. First statement is FALSE(0) Then check the second statement
     c. Both the statement are FALSE(0) the Output is False(0)

C. Logical NOT (not)

   A     not(A)
   0       1
   1       0

In [80]:
x  = 6
print(x < 5 and x<10)

False


In [81]:
print(x<5 or x<4)

False


In [82]:
print(not(x==5))

True


In [83]:
print(not(x<5 and x<10)) #Here first it will check inner condition.

True


### 5. Bitwise Operator

The two operands values are processed bit by bit by the bitwise operators.

#### Example:-

if a = 7
   b = 6
then, binary (a) = 0111
    binary (b) = 0110

hence, a & b = 0011 (Binary and)
      a | b = 0111  (Binary or)
             a ^ b = 0100 (Binary xor)
       ~ a = 1000   (negation)

## Python if.....else
###### Python supports the usual logical conditions from mathematics:

     Equals: a == b
     Not Equals: a != b
     Less than: a < b
     Less than or equal to: a <= b
     Greater than: a > b
     Greater than or equal to: a >= b
###### These conditions can be used in several ways, most commonly in "if statements" and loops.

###### An "if statement" is written by using the if keyword.


In [84]:
a = int(input("Enter the first number :"))
b = int(input("Enter the second number :"))
c = int(input("Enter the third number :"))
if a > b:
    print("The King won")

elif b > c:
    print("The Queen won")
else:
    print("Sorry Babes So Jao Ab")

The King won


### Python Loops
###### Python has two primitive loop commands:

     1. while loops
     2. for loops

Note: In while loop remember to increment i, or else the loop will continue forever.

With the while loop we can execute a set of statements as long as a condition is true.

#### The break Statement
With the break statement we can stop the loop even if the while condition is true:

#### The continue Statement
With the continue statement we can stop the current iteration, and continue with the next:

### The else Statement
With the else statement we can run a block of code once when the condition no longer is true:

In [85]:
i = 1
while i < 6:
  print(i)
  i += 1

1
2
3
4
5


In [86]:
i = 1
while i < 6:
  print(i)
  if i == 3:
    break
  i += 1

1
2
3


In [87]:
i = 0
while i < 6:
  i += 1
  if i == 3:
    continue
  print(i)

1
2
4
5
6


In [88]:
i = 1
while i < 6:
  print(i)
  i += 1
else:
  print("i is no longer less than 6")

1
2
3
4
5
i is no longer less than 6


### Python For Loops
1. A for loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string).
2. This is less like the for keyword in other programming languages, and works more like an iterator method as found in other object-orientated programming languages.
3. With the for loop we can execute a set of statements, once for each item in a list, tuple, set etc.




In [89]:
#Looping Through a String
#Even strings are iterable objects, they contain a sequence of characters:

for x in "banana":
  print(x)

b
a
n
a
n
a


In [90]:
#The break Statement
#With the break statement we can stop the loop before it has looped through all the items:
#Exit the loop when x is "banana":

fruits = ["apple", "banana", "cherry"]
for x in fruits:
  print(x)
  if x == "banana":
    break

apple
banana


In [91]:
#Exit the loop when x is "banana", but this time the break comes before the print:

fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    break
  print(x)

apple


In [92]:
#The continue Statement
#With the continue statement we can stop the current iteration of the loop, and continue with the next:

fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    continue
  print(x)

apple
cherry


### The range() Function
1. To loop through a set of code a specified number of times, we can use the range() function,
2. The range() function returns a sequence of numbers, starting from 0 by default, and increments by 1 (by default), and ends at a specified number.


Note: The else block will NOT be executed if the loop is stopped by a break statement.

In [93]:
#Using the range() function:

for x in range(6):
  print(x)
print("--------------------------------")
#Note that range(6) is not the values of 0 to 6, but the values 0 to 5.
#The range() function defaults to 0 as a starting value, however it is possible to specify the starting value by adding a parameter: range(2, 6), which means values from 2 to 6 (but not including 6):

for x in range(2, 6):
  print(x)

0
1
2
3
4
5
--------------------------------
2
3
4
5


In [94]:
#The range() function defaults to increment the sequence by 1, however it is possible to specify the increment value by adding a third parameter: range(2, 30, 3):

for x in range(2, 30, 3):
  print(x)

2
5
8
11
14
17
20
23
26
29


In [95]:
#Else in For Loop
#The else keyword in a for loop specifies a block of code to be executed when the loop is finished:


for x in range(6):
  print(x)
else:
  print("Finally finished!")

0
1
2
3
4
5
Finally finished!


In [96]:
#Break the loop when x is 3, and see what happens with the else block:

for x in range(6):
  if x == 3: break
  print(x)
else:
  print("Finally finished!")

0
1
2


In [97]:
for x in range(6):
    print(x)
    if x == 3: break
else:
  print("Finally finished!")

0
1
2
3


In [98]:
#Nested Loops
#A nested loop is a loop inside a loop.

#The "inner loop" will be executed one time for each iteration of the "outer loop":


adj = ["red", "big", "tasty"]
fruits = ["apple", "banana", "cherry"]

for x in adj:
  for y in fruits:
    print(x, y)

red apple
red banana
red cherry
big apple
big banana
big cherry
tasty apple
tasty banana
tasty cherry


In [99]:
#The pass Statement
#for loops cannot be empty, but if you for some reason have a for loop with no content, put in the pass statement to avoid getting an error.


for x in [0, 1, 2]:
  pass